storage/ksql/src/main/java/io/apicurio/registry/storage/impl/ksql/KafkaSqlRegistryStorage.java

```diff
@@ -0,0 +1,597 @@
+    }
+
+    @Override
+    public void updateArtifactState(String artifactId, ArtifactState state, Integer version) {
+        ArtifactVersionMetaDataDto metadata = sqlStorage.getArtifactVersionMetaData(artifactId, version);
+        updateArtifactState(metadata.getState(), artifactId, version, state);
+    }
+
+    @Override
+    public CompletionStage<ArtifactMetaDataDto> createArtifact(String artifactId, ArtifactType artifactType, ContentHandle content) throws ArtifactAlreadyExistsException, RegistryStorageException {
+
+        try {
+            sqlStorage.getArtifactMetaData(artifactId);
+            throw new ArtifactAlreadyExistsException(artifactId);
+        } catch (ArtifactNotFoundException e) {
+            //ignored
+            //artifact does not exist, we can create it
+        }
+
+        return submitter
+                .submitArtifact(Str.ActionType.CREATE, artifactId, -1, artifactType, content.bytes())
+                .thenCompose(reqId -> {
+                    return (CompletionStage<ArtifactMetaDataDto>) coordinator.waitForResponse(reqId);
+                });
+    }
+
+    @Override
+    public CompletionStage<ArtifactMetaDataDto> createArtifactWithMetadata(String artifactId, ArtifactType artifactType, 
+            ContentHandle content, EditableArtifactMetaDataDto metaData) throws ArtifactAlreadyExistsException, RegistryStorageException {
+        return createArtifact(artifactId, artifactType, content)
+            .thenCompose(amdd -> submitter.submitMetadata(Str.ActionType.UPDATE, artifactId, -1, metaData.getName(), metaData.getDescription(), metaData.getLabels(), metaData.getProperties())
+                .thenApply(v -> DtoUtil.setEditableMetaDataInArtifact(amdd, metaData)));
+    }
+
+    @Override
+    public SortedSet<Long> deleteArtifact(String artifactId) throws ArtifactNotFoundException, RegistryStorageException {
+
+        //to verify artifact exists
+        //TODO implement a low level storage api that provides methods like, exists, ...
+        sqlStorage.getArtifactMetaData(artifactId);
+
+        UUID reqId = ConcurrentUtil.get(submitter.submitArtifact(Str.ActionType.DELETE, artifactId, -1, null, null));
+        SortedSet<Long> versionIds = (SortedSet<Long>) coordinator.waitForResponse(reqId);
+
+        return versionIds;
+    }
+
+    @Override
+    public StoredArtifact getArtifact(String artifactId) throws ArtifactNotFoundException, RegistryStorageException {
+        return sqlStorage.getArtifact(artifactId);
+    }
+
+    @Override
+    public CompletionStage<ArtifactMetaDataDto> updateArtifact(String artifactId, ArtifactType artifactType, ContentHandle content) throws ArtifactNotFoundException, RegistryStorageException {
+
+        try {
+            sqlStorage.getArtifactMetaData(artifactId);
+        } catch (ArtifactNotFoundException e) {
+            throw e;
+        }
+
+        return submitter
```

Comment:
I don't see how the artifact version is being handled in this storage. I know that the -1 is what the streams storage is sending at this point, but then the `StreamsTopologyProvider` updates that version appropriately. I don't see something that replaces that functionality here.

Reply:
The version is created when the SQL layer creates the row in the `versions` table.  I believe the try-catch is purely a fail-fast check for when the artifact doesn't exist.

The way this will work is that a Kafka message with the artifactId and a "create version" action (and the content) will be published.  Then all replicas will consume that message and perform the action, which is to create a new version for the given artifactId.  At that point the version will be generated and communicated back to the original thread via the `coordinator`.

Does that explanation make sense?

Reply:
Yes, it does. Thanks, now I see the query and its usage in the `CommonSqlStatements`.